### PR TITLE
TypeScript upgrade (and more)

### PR DESCRIPTION
**Pull request by jcbhmr**
<div align="center">

**🚧 Under construction! 👷‍♂️**

</div>

I don't really know how to break this apart into constituent PRs. If there's an easy way, let me know! I figured it'd just be easier if I made all my fancy interconnected and interdependent changes in one go and then had you review them all and make a bunch of nitpicks before merging. I've also tried (with some success) to add some helpful "Why?" descriptions to the commits I've made.

This PR would...
- [x] Use the official GitHub .gitignore for Node.js
- [x] Convert the source files to TypeScript
- [ ] Add more inline code comments
- [x] Add badge icons to the readme
- [ ] Add more detailed getting started instructions and examples to the readme
- [ ] Provide a more streamlined demo link (direct GitHub Pages, no sub demo/* directory)
- [x] Deploy to npm using GitHub Actions (**⚠️ Needs `NPM_TOKEN` secret!** -- see below)
- [x] Revamp package.json to use more modern conventions
- [x] Change the build tool to Vite
- [ ] ~~Change the test tool to Vitest~~
- [x] Change the test tool to Playwright

<details>
  <summary>For maintainer: Adding an NPM_TOKEN secret</summary>

This is an abbreviated guide based on:
- https://docs.github.com/en/actions/publishing-packages/publishing-nodejs-packages
- https://dev.to/astagi/publish-to-npm-using-github-actions-23fn

**BUT!** Instead of using a complete repository token accessible to ANY GitHub Action (even simple stuff like linting), we can set it so that the token is only available in a specific environment. This means you can add special reviewers, special rules, etc. for code that touches this (I think? This is my first time playing with environments).

📚 https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment

1. Get an npm access token for Automation
2. Go to GitHub settings tab for your repo
3. Do something... (WIP)

</details>

### CODE REVIEW

**Comment by jcbhmr** [comment]
Note to self: Here's some of the links for the HTML `<style scoped>` specs:
- MDN page from late in the lifecycle mid 2017 (it's deprecated at this point): https://web.archive.org/web/20170702055616/https://developer.mozilla.org/en-US/docs/Web/HTML/Element/style
- GitHub issue to remove it from the HTML spec (Jan 2016): https://github.com/whatwg/html/issues/552

**Comment by jcbhmr** [comment]
@samthor Looking for thoughts on my changes to the readme. Is this a sort of style that's acceptable? Would you prefer more text? More code examples? More links?

https://github.com/jcbhmr/samthor-scoped#readme

**Comment by jcbhmr** [comment]
The GitHub test action works! Here's an example of the log: https://github.com/jcbhmr/samthor-scoped/actions/runs/3952666028/jobs/6768021281

image

**Comment by jcbhmr** [comment]
I made a somewhat drastic change where I completely axed the demo/ dir in favor of the test/ dir. I think that a future step might be to use a CodeSandbox playground instead of a static GitHub Pages site.

https://5ocrf8.csb.app/



Advantages of using a CodeSandbox site:
- User can edit the CSS and see live feedback
- Can see code and result side-by-side

Advantages of using GitHub Pages
- Hosted on GitHub (no relying on some other user to not delete the CodeSandbox project)

File: test/selectors.html [hunk 1]
```diff
@@ -0,0 +1,46 @@
+<!DOCTYPE html>
+<script type="importmap">
+  {
+    "imports": {
```
Comment (samthor) [inline test/selectors.html:4]: It's the future.

**Comment by samthor** [comment]
I appreciate all the help. Is there a 'done' point here—mark it as not draft?

**Comment by jcbhmr** [comment]
Yep, I still want to tidy up some of the TS code to make it _actually_ ts and maybe some other stuff.

If you know of somehow to take specific commits and pull/split them onto their own branch, I can individually split out each thing into its own PR. I just am not fluent in git commands to know how to do that.

**Comment by jcbhmr** [comment]
Back at it again, I'm going to try and splinter this PR into multiple different focused PRs. I have been putting off diving into how to do this in git since it seemed complicated at first glance. (You may have noticed that I remarked about in this PR's description and once in the comments)

I'm going to try to use these (below) resources and split this PR. We'll see how far I get.

- https://gist.github.com/loilo/930f141d9acf89e9e734ffa042acd750
- https://gist.github.com/loilo/930f141d9acf89e9e734ffa042acd750?permalink_comment_id=3136196#gistcomment-3136196
- https://medium.com/@groksrc/protip-how-to-split-large-branches-into-small-pull-requests-81d607660c05

This comment is both for future me and to keep @samthor in-the-loop.

**Comment by jcbhmr** [comment]
I think I've figured out how I'm going to change this into multiple PRs:

1. Create a new branch from the `main` that I've been working on called "big-branch"
2. Force revert main to its original state (jcbhmr's main == samthor's main)
3. Create a branch called "use-vite-and-ts"
4. Copy-and-paste the files from "big-branch" that fit the theme of "Vite and TS" to that new branch
5. Open a PR with "use-vite-and-ts"
6. Repeat for each self-contained change

We'll see how far this gets me before I need to start basing PRs on top of PRs! 😉

**Pull request closed without merging**